src/modules/audio/audio.c

```diff
@@ -19,7 +19,18 @@ static const ma_format miniAudioFormatFromLovr[] = {
 #define OUTPUT_FORMAT SAMPLE_F32
 #define OUTPUT_CHANNELS 2
 #define CAPTURE_CHANNELS 1
-#define CAPTURE_BUFFER_SIZE ((int)(LOVR_AUDIO_SAMPLE_RATE * 1.0))
+
+#define CALLBACK_PERIODS 3
+#define PERIOD_LENGTH 128
+#define CALLBACK_LENGTH (PERIOD_LENGTH*CALLBACK_PERIODS)
+
+//#define LOVR_DEBUG_AUDIOTAP
+#ifdef LOVR_DEBUG_AUDIOTAP
+// To get a record of what the audio callback is playing, define LOVR_DEBUG_AUDIOTAP,
+// after running look in the lovr save directory for lovrDebugAudio.raw,
+// and open as raw 32-bit stereo floats (Audacity can do this, or Amadeus on Mac)
+#include "filesystem/filesystem.h"
+#endif
```

Comment:
I am a little weary about maintaining this.  It seems prone to rot since it will be infrequently used, and adds a inter-module dependency which I try to avoid.

I would be more comfortable turning this into a real feature like "loopbacks".  miniaudio has real loopback devices but they're windows only.  It could maybe be emulated like this:

- In the playback callback, if loopback is enabled write everything to a ringbuffer/stream-SoundData.  Expose that to Lua so people can do whatever they want with the contents (write to file, send over network, calculate amplitude, etc.).  Loopback could be treated as enabled if Lua has set a loopback destination buffer (`lovr.audio.setLoopback(stream|nil)`, or whatever it's named).

Reply:
All I know is I wind up hacking this in almost every time I need to work on lovr audio.

Would you be more comfortable if there were an explicit comment saying "this is for debug convenience but not maintained"? The loopback feature sounds super cool but I don't know how long it would take us to develop it.

(In one of my portaudio branches I implemented an ad hoc loopback feature like this and used it to display the waveforms in realtime-- it was pretty nice!)

Reply:
(To be clear if you're asking me to remove it, I'll remove it)

Reply:
I can add the loopback feature before merging miniaudio-new